# Universidad de las Fuerzas Armadas ESPE

## Andrés Cosios Pineda

## Estudio de Caso

### 1. En el libro Numerical Mathematics and Computing de Cheney y Kincaid, observe el ejercicio 4, literal a de la página 288 (el libro se encuentra en el enlace proporcionado en la sección Sitios).
### Use procedure Tri to solve the following system of 100 equations. Compare the numerical solution to the obvious exact solution.
### $x_1+0.5x_2=1.5$
### $0.5x_{i-1}+x_i+0.5x_{i+1}=2.0$
### $0.5x_{99}+x_{100}=1.5$
###        $ (2<=i<=99)$

### 2. Utilice el siguiente código para resolver el ejercicio (utilizando un sistema de cien ecuaciones y cien incógnitas). El código es del mismo libro. 
![](metodos.jpg)


In [4]:
import numpy as np
# creamos las matrices
# vector diagonal
d=np.array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
            1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
            1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
            1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
            1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
            1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
            1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
            1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
            1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
            1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
# vector Subdiagonal
a=np.array([0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5])
# Vector superdiagonal
c=np.array([0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5,
            0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5])
# vector de terminos independientes
b=np.array([1.5, 2., 2., 2., 2., 2., 2., 2., 2., 2.,
           2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
           2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
           2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
           2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
           2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
           2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
           2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
           2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
           2., 2., 2., 2., 2., 2., 2., 2., 2., 1.5])
n=100

def tridiag(n,d,a,c,b):
    x=np.zeros(n)
    for i in range (1, n-1):
        mult=a[i-1]/d[i-1]
        d[i]-=mult*c[i-1]
        b[i]-=mult*b[i-1]
    print 
    print(d)
    x[n-1]=b[n-1]/d[n-1]
    for i in range(n-2,-1,-1):
        x[i]=(b[i]-c[i]*x[i+1])/d[i]
    return x
tridiag(n,d,a,c,b)

[1.         0.75       0.66666667 0.625      0.6        0.58333333
 0.57142857 0.5625     0.55555556 0.55       0.54545455 0.54166667
 0.53846154 0.53571429 0.53333333 0.53125    0.52941176 0.52777778
 0.52631579 0.525      0.52380952 0.52272727 0.52173913 0.52083333
 0.52       0.51923077 0.51851852 0.51785714 0.51724138 0.51666667
 0.51612903 0.515625   0.51515152 0.51470588 0.51428571 0.51388889
 0.51351351 0.51315789 0.51282051 0.5125     0.51219512 0.51190476
 0.51162791 0.51136364 0.51111111 0.51086957 0.5106383  0.51041667
 0.51020408 0.51       0.50980392 0.50961538 0.50943396 0.50925926
 0.50909091 0.50892857 0.50877193 0.50862069 0.50847458 0.50833333
 0.50819672 0.50806452 0.50793651 0.5078125  0.50769231 0.50757576
 0.50746269 0.50735294 0.50724638 0.50714286 0.50704225 0.50694444
 0.50684932 0.50675676 0.50666667 0.50657895 0.50649351 0.50641026
 0.50632911 0.50625    0.50617284 0.50609756 0.5060241  0.50595238
 0.50588235 0.50581395 0.50574713 0.50568182 0.50561798 0.5055

array([0.995, 1.01 , 0.985, 1.02 , 0.975, 1.03 , 0.965, 1.04 , 0.955,
       1.05 , 0.945, 1.06 , 0.935, 1.07 , 0.925, 1.08 , 0.915, 1.09 ,
       0.905, 1.1  , 0.895, 1.11 , 0.885, 1.12 , 0.875, 1.13 , 0.865,
       1.14 , 0.855, 1.15 , 0.845, 1.16 , 0.835, 1.17 , 0.825, 1.18 ,
       0.815, 1.19 , 0.805, 1.2  , 0.795, 1.21 , 0.785, 1.22 , 0.775,
       1.23 , 0.765, 1.24 , 0.755, 1.25 , 0.745, 1.26 , 0.735, 1.27 ,
       0.725, 1.28 , 0.715, 1.29 , 0.705, 1.3  , 0.695, 1.31 , 0.685,
       1.32 , 0.675, 1.33 , 0.665, 1.34 , 0.655, 1.35 , 0.645, 1.36 ,
       0.635, 1.37 , 0.625, 1.38 , 0.615, 1.39 , 0.605, 1.4  , 0.595,
       1.41 , 0.585, 1.42 , 0.575, 1.43 , 0.565, 1.44 , 0.555, 1.45 ,
       0.545, 1.46 , 0.535, 1.47 , 0.525, 1.48 , 0.515, 1.49 , 0.505,
       1.5  ])

### 3. Resuelva el sistema de ecuaciones de manera manual y compare los resultados. Concluya lo que sucede en este caso. Puede observar la solución al ejercicio en el libro.

### tenemos que una matriz tridiagonal es de la siguiente forma:

![](trid.jpg)
### vamos a despejar dos ecuaciones para determinar el valor de la matriz diagonal $d_1$
### $x_i+0.5x_{i+1}=1.5$
#### Despejamos  $x_1$ en la primera ecuación
### $x_i=1.5-0.5x_{i+1}$
#### Despejamos  $x_1$ en la segunda ecuación
### $0.5x_{i-1}+x_i+0.5x_{i+1}=2.0$
### $x_i=2-0.5x{i+1}-0.5x{i-1}$
#### Igualamos las ecuaciones
### $1.5-0.5x_{i+1}=2-0.5x{i+1}-0.5x{i-1}$
#### Obtenemos que $x_1=1$, por tanto $d=1$
### Determinamos el termino independiente de la primera interaccion i=1
### $\widehat{d}=\frac{b_1}{d_1}=\frac{1.5}{1}=1.5$
### Realizaremos dos iteracciones mas
### a partir de estas itercacciones utilizaremos las siguientes formulas
### $ \widehat{d_i}=d_i-\frac{a_{i-1}}{d_{i-1}}*(c_{i-1})$
### $ \widehat{b_i}=b_i-\frac{a_{i-1}}{d_{i-1}}*(b_{i-1})$
### con: $i=2$
### $a_1=0.5, d_1=1, c_1=0.5, b_1=1.5$
### $ \widehat{d_2}=d_2-\frac{a_1}{d_1}*(c_1)=1-\frac{0.5}{1}*0.5=0.75$
### $ \widehat{b_2}=b_2-\frac{a_1}{d_1}*(b_1)=2-\frac{0.5}{1}*1.5=1.25$
### con: $i=3$
### $a_2=0.5, d_2=0.75, c_2=0.5, b_2=2$
### $ \widehat{d_3}=d_3-\frac{a_2}{d_2}*(c_2)=1-\frac{0.5}{0.75}*0.5=0.666$
### $ \widehat{b_3}=b_3-\frac{a_2}{d_2}*(b_2)=2-\frac{0.5}{0.75}*1.25=1.166$

### Concluciones

#### Para poder comparar el funcionamiento del algoritmo indicado en la actividad se procedio con la realizacion manual de tres iteracciones las cuales coinciden con los resultados obtenidos en el programa desarollado en Python observando que con la implemntacion de unas lineas de codigo podemos obtener todas las iteracciones posibles simplemente modificando los parametros del mismo.

### Link Github
#### https://github.com/aacosios/Estudio-de-Caso-1.git